src/modules/bulletproofs/rangeproof_impl.h

```diff
@@ -209,31 +209,45 @@ static int secp256k1_bulletproof_rangeproof_verify_impl(const secp256k1_ecmult_c
         size_t j;
 
         /* Commit to all input data: min value, pedersen commit, asset generator, extra_commit */
-        if (min_value != NULL && min_value[i] != NULL) {
-            secp256k1_sha256_initialize(&sha256);
-            secp256k1_sha256_write(&sha256, commit, 32);
-            for (j = 0; j < n_commits; j++) {
-                unsigned char vbuf[8];
-                vbuf[0] = min_value[i][j];
-                vbuf[1] = min_value[i][j] >> 8;
-                vbuf[2] = min_value[i][j] >> 16;
-                vbuf[3] = min_value[i][j] >> 24;
-                vbuf[4] = min_value[i][j] >> 32;
-                vbuf[5] = min_value[i][j] >> 40;
-                vbuf[6] = min_value[i][j] >> 48;
-                vbuf[7] = min_value[i][j] >> 56;
-                secp256k1_sha256_write(&sha256, vbuf, 8);
-            }
-            secp256k1_sha256_finalize(&sha256, commit);
-        }
+        secp256k1_sha256_initialize(&sha256);
+        secp256k1_sha256_write(&sha256, commit, 32);
         for (j = 0; j < n_commits; j++) {
-            secp256k1_bulletproof_update_commit(commit, &commitp[i][j], &value_gen[i]);
+            unsigned char vbuf[8];
+            uint64_t mv;
+            if (min_value != NULL && min_value[i] != NULL) {
+                mv = min_value[i][j];
+            } else {
+                mv = 0;
+            }
+
+            vbuf[0] = mv;
+            vbuf[1] = mv >> 8;
+            vbuf[2] = mv >> 16;
+            vbuf[3] = mv >> 24;
+            vbuf[4] = mv >> 32;
+            vbuf[5] = mv >> 40;
+            vbuf[6] = mv >> 48;
+            vbuf[7] = mv >> 56;
+            secp256k1_sha256_write(&sha256, vbuf, 8);
         }
+
         if (extra_commit != NULL && extra_commit[i] != NULL) {
-            secp256k1_sha256_initialize(&sha256);
-            secp256k1_sha256_write(&sha256, commit, 32);
+            unsigned char len_buf[4];
+            len_buf[0] = extra_commit_len[i];
+            len_buf[1] = extra_commit_len[i] >> 8;
+            len_buf[2] = extra_commit_len[i] >> 16;
+            len_buf[3] = extra_commit_len[i] >> 24;
+
+            secp256k1_sha256_write(&sha256, len_buf, 4);
```

Comment:
The length of this buffer is (was?) not my concern. The thing is that there's ambiguity between the cases 
 * `n_commits` = n and an extra commitment (only with length 4 bytes after your commit)
 * `n_commits` = n+1  and no extra commitment

for some n.

I think the simplest way to address this is to write the value of `n_commits` to the hash before the loop. Then we don't need to write the length of the extra commitment to the hash, it's just all the data after the last commitment. (Again, I have no idea of the consensus rules but we need to make sure that a "NULL" extra commitment is not malleable to an extra commitment with an empty string, similar to my comment about the min value above).

Reply:
In the consensus rules the `extra_commitment` will always be the scriptpubkey, or something like that ... once again it is determined by consensus and not ever serialized as part of the proof.